contributors/design-proposals/sig-storage/volume-topology-scheduling.md

```diff
@@ -0,0 +1,580 @@
+when it schedules the pod to a node.  A new scheduling predicate is still
+needed to filter and match the PVs (but not actually bind).
+
+Major downsides of this approach include:
+* Requires PV controller to change its sync loop to operate on pods, in order
+to handle the multiple PVCs in a pod scenario.  This is a potentially big change
+that would be hard to keep separate and feature-gated from the current PV logic.
+* Needs new coordination between PV controller and kubelet to reject a pod if
+binding fails to send the pod back to scheduler.
+* Initial filtering (matching) and binding are done asynchronously so there is a
+higher chance of binding failures.  As an example, we could match PV1 for both
+pod1 and pod2, but only pod1 will be able to bind successfully, and pod2 has to
+be sent back to scheduling.
+* More race conditions and higher chance of binding failure because of 3
+different components operating independently (scheduler, PV controller, kubelet).
+
+##### Move PVC binding to kubelet
+Looking into the future, with the potential for NUMA-aware scheduling, you could
+have a sub-scheduler on each node to handle the pod scheduling within a node.  It
+could make sense to have the volume binding as part of this sub-scheduler, to make
+sure that the volume selected will have NUMA affinity with the rest of the
+resources that the pod requested.
+
+However, there are potential security concerns because kubelet would need to see
+unbound PVs in order to bind them.  For local storage, the PVs could be restricted
+to just that node, but for zonal storage, it could see all the PVs in that zone.
+
+In addition, the sub-scheduler is just a thought at this point, and there are no
+concrete proposals in this area yet.
+
+### Binding multiple PVCs in one transaction
+For the alpha phase, this is not required.  Since the scheduler is serialized, a
+partial binding failure should be a rare occurrence.  Manual recovery will need
+to be done for those rare failures.
+
+### Recovering from kubelet rejection of pod
+Again, for alpha phase, manual intervention will be required.  For beta, we’ll
+have to consider some sort of rollback mechanism for more robustness.  Perhaps
+this rollback mechanism can also be used for binding multiple PVCs in one
+transaction.
+
+### Making dynamic provisioning topology aware
+TODO (beta): Design details
+
+For alpha, we are not focusing on this use case.  But it should be able to
+follow the new workflow closely with some modifications.
+* The FindUnboundPVCs predicate function needs to get provisionable capacity per
+topology dimension from the provisioner somehow.
+* The PrioritizeUnboundPVCs priority function can add a new priority score factor
+based on available capacity per node.
+* The BindUnboundPVCs bind function needs to pass in the node to the provisioner.
+The internal and external provisioning APIs need to be updated to take in a node
+parameter.
+
+
```

Comment:
This is definitely going to make it harder to run your own scheduler, and I think this is also one of the reasons why the scheduling sig added the scheduler extender interface, so that you can still use the default scheduler + your custom extensions.  In general, the scheduling team has been adding many advanced scheduling features to the default scheduler that it's becoming harder to run your own already.

If you don't use the default scheduler, and you use persistent volumes, then you'll have to pull this code and integrate it into your custom scheduler.  Of course, we will have to give notice well in advance.

Reply:
I thought more about it, if two pods use the same unbound PVC and two different schedulers bind this PVC to a different PV (i.e. one saves `pv1.spec.claimRef = pvc` and the second one `pv2.spec.claimRef = pvc`), PV controller is smart enough to recover from this conflict by un-binding one of these PVs, i.e. it clears say `pv1.spec.claimRef=nil` (and deletes it when necessary).

If both schedulers implement step `"9. **NEW:** If PVC binding or provisioning is required, we do NOT AssumePod ... then **wait until the PVCs are bound successfully** or encounter failure"` then the PVC gets eventually bound to one of these PVs. Not necessarily the one that a particular scheduler has chosen, but that should not be a problem, right?

So, we're fine as long as both schedulers implement the rules you specified. They may generate some noise by provisioning an unnecessary PV or do extra binding, but PV controller should clean up this mess.

Reply:
Thanks for the analysis.  It should be fine, but the only potential problem is if the 2nd scheduler bound the PVC in a way that cannot satisfy the all the predicates in the 1st scheduler.  In general, we should only recommend that one scheduler handle PVC binding.

Can @kubernetes/sig-scheduling-pr-reviews comment on the practice of running two schedulers against the same Pod?  Is that a supported configuration, or do we only support a model where:
- You run your own scheduler by itself
- You run default scheduler + your custom extensions